index.js

```diff
@@ -975,25 +977,48 @@ const _bindWindow = (window, newWindowCb) => {
         }
       }
     }
+    window.requestAnimationFrame(
+      () => {
+        if (args.performance) {
+          const now = Date.now();
+          const diff = now - timestamps.last;
+          timestamps.user += diff;
+          timestamps.total += diff;
+          timestamps.last = now;
+        }
+        _blit(true);
+        if (args.performance) {
+          const now = Date.now();
+          const diff = now - timestamps.last;
+          timestamps.submit += diff;
+          timestamps.total += diff;
+          timestamps.last = now;
+        }
+      },
+      1);
```

Comment:
So overall this seems like the frame flow is:

1. Update matrices
2. Blit do HMD
3. Run user rafs
4. Bit to window

This doesn't seem right to me. Should the user code always run first? Otherwise as mentioned the HMD blit will be one frame behind.

Reply:
The user code can't run first unless you figure out some way of deciding whether the HMD needs to blit its split screen to the window.

Right now the HMD blits to the window, then the window can overwrite that blit with a different render. If you reverse it, the HMD blit will stomp on the window render.

I'm not sure there's a good way in general to detect whether a site is trying to do separate renders for window vs device.

Ideas are welcome, but it seems like a tough problem.

Reply:
> I'm not sure there's a good way in general to detect whether a site is trying to do separate renders for window vs device.

Well at this point in the code we already know that don't we? We know whether the user has requested device or window rafs based on which they called. This was previously (and can still) be flagged at raf registration time.

> Right now the HMD blits to the window, then the window can overwrite that blit with a different render. If you reverse it, the HMD blit will stomp on the window render.

So it seems the order should be:

1. Update matrices
1. Run device rafs
1. If there were device rafs, blit to device AND back to window, otherwise no blits
1. Run window rafs
1. If there were window rafs, blit to window

Reply:
And how do we determine whether there were device rafs? :) That was the functionality you removed.

Reply:
Indeed, but that can easily be put back.

Reply:
Actually I think this is a bit more complicated since you probably don't want to destructively blit in `3` if there was no context update. That's what `gl.isDirty()` is for (it's tracked automatically at the C++ layer).

Reply:
Yes, that's what I meant by "Certain prios should only run if certain other prios run." Not so crazy after all!

Unfortunately splitting it up by prio makes that logic difficult. Ideas?

Reply:
I don't think any of the above code should use raf itself. Its job is to execute user rafs, but it shouldn't need to use them. That gets rid of the problem of dictating over raf prios.

Just straight ifs and function calls should do it I think.

Reply:
If the blits don't become RAFs, what was the point of the prio refactor with sorting? I thought blits were supposed to be RAFs?

Reply:
The prio refactor was to support the needs of `exokit-home`, which is for all purposes effectively just a user site. So it's allowed to dabble in and leverage the raf sorting feature.

I don't think the same goes for the core -- it's not a user site and it has enough context to not need to resort on raf sorting.